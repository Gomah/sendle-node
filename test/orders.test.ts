import { describe, it, expect } from 'vitest';
import { client } from './helpers';
import * as dateFns from 'date-fns';
let orderId: string;
let trackingId: string;

describe('Orders', () => {
  it('should create a domestic order', async () => {
    const order = await client.orders.create({
      first_mile_option: 'pickup',
      description: 'Test',
      weight: {
        value: '1',
        units: 'kg',
      },
      customer_reference: '1337',
      metadata: {
        userId: 100,
      },
      sender: {
        contact: {
          name: 'Lex Luthor',
        },
        address: {
          address_line1: '123 Gotham Ln',
          suburb: 'Sydney',
          state_name: 'NSW',
          postcode: '2000',
          country: 'Australia',
        },
      },
      receiver: {
        instructions: 'Signature on Delivery',
        contact: {
          name: 'Clark Kent',
          email: 'clarkissuper@dailyplanet.xyz',
          company: 'Daily Planet',
        },
        address: {
          address_line1: '80 Wentworth Park Road',
          suburb: 'Darwin',
          state_name: 'NT',
          postcode: '0800',
          country: 'Australia',
        },
      },
    });

    orderId = order.order_id;
    trackingId = order.sendle_reference;

    expect(order).toMatchSnapshot({
      order_id: expect.any(String),
      order_url: expect.any(String),
      sendle_reference: expect.any(String),
      tracking_url: expect.any(String),
      labels: expect.any(Array),
      scheduling: {
        pickup_date: expect.any(String),
        estimated_delivery_date_minimum: expect.any(String),
        estimated_delivery_date_maximum: expect.any(String),
      },
      sender: {
        contact: {
          email: expect.any(String),
          sendle_id: expect.any(String),
        },
      },
      price: {
        gross: {
          amount: expect.any(Number),
        },
        net: {
          amount: expect.any(Number),
        },
        tax: {
          amount: expect.any(Number),
        },
      },
      price_breakdown: {
        base: {
          amount: expect.any(Number),
        },
        base_tax: {
          amount: expect.any(Number),
        },
        fuel_surcharge: {
          amount: expect.any(Number),
        },
        fuel_surcharge_tax: {
          amount: expect.any(Number),
        },
      },
      tax_breakdown: {
        gst: {
          amount: expect.any(Number),
        },
      },
    });
  });

  it('should create an international order', async () => {
    const order = await client.orders.create({
      pickup_date: dateFns.format(
        dateFns.add(new Date(), {
          days: 3,
        }),
        'yyyy-MM-dd'
      ),
      description: 'Kryptonite',
      weight: {
        value: '1',
        units: 'kg',
      },
      volume: {
        value: '0.01',
        units: 'm3',
      },
      customer_reference: 'SupBdayPressie',
      metadata: {
        your_data: 'XYZ123',
      },
      sender: {
        contact: {
          name: 'Lex Luthor',
          phone: '0412 345 678',
        },
        address: {
          address_line1: '123 Gotham Ln',
          suburb: 'Sydney',
          state_name: 'NSW',
          postcode: '2000',
          country: 'Australia',
        },
        instructions: 'Knock loudly',
      },
      receiver: {
        contact: {
          name: 'Clark Kent',
          email: 'clarkissuper@dailyplanet.xyz',
        },
        address: {
          address_line1: '445 Mount Eden Road',
          suburb: 'Auckland',
          postcode: '2025',
          country: 'New Zealand',
        },
        instructions: 'Give directly to Clark',
      },
      parcel_contents: [
        {
          description: 'T-shirt',
          value: '20.00',
          country_of_origin: 'China',
          hs_code: '610910',
        },
      ],
    });

    expect(order).toMatchSnapshot({
      order_id: expect.any(String),
      order_url: expect.any(String),
      sendle_reference: expect.any(String),
      tracking_url: expect.any(String),
      labels: expect.any(Array),
      scheduling: {
        pickup_date: expect.any(String),
        estimated_delivery_date_minimum: expect.any(String),
        estimated_delivery_date_maximum: expect.any(String),
      },
      sender: {
        contact: {
          email: expect.any(String),
          sendle_id: expect.any(String),
        },
      },
      price: {
        gross: {
          amount: expect.any(Number),
        },
        net: {
          amount: expect.any(Number),
        },
        tax: {
          amount: expect.any(Number),
        },
      },
      price_breakdown: {
        base: {
          amount: expect.any(Number),
        },
        base_tax: {
          amount: expect.any(Number),
        },
        fuel_surcharge: {
          amount: expect.any(Number),
        },
        fuel_surcharge_tax: {
          amount: expect.any(Number),
        },
      },
    });
  });

  it('should return tracking for my order', async () => {
    const tracking = await client.tracking(trackingId);

    expect(tracking).toMatchSnapshot({
      scheduling: {
        pickup_date: expect.any(String),
        estimated_delivery_date_minimum: expect.any(String),
        estimated_delivery_date_maximum: expect.any(String),
      },

      status: {
        description: expect.any(String),
        last_changed_at: expect.any(String),
      },
    });
  });

  it('should return A4 label for my order', async () => {
    const label = await client.labels.get({ orderId });

    expect(label).toContain('%PDF-1.4\n');
  });

  it('should cancel an order', async () => {
    const order = await client.orders.cancel(orderId);

    expect(order).toMatchSnapshot({
      order_id: expect.any(String),
      cancelled_at: expect.any(String),
      cancellation_message: expect.any(String),
      order_url: expect.any(String),
      sendle_reference: expect.any(String),
      tracking_url: expect.any(String),
    });
  });
});
