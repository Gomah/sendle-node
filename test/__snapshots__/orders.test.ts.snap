// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Orders > should cancel an order 1`] = `
{
  "cancellation_message": Any<String>,
  "cancelled_at": Any<String>,
  "customer_reference": "1337",
  "metadata": {
    "userId": 100,
  },
  "order_id": Any<String>,
  "order_url": Any<String>,
  "sendle_reference": Any<String>,
  "state": "Cancelled",
  "tracking_url": Any<String>,
}
`;

exports[`Orders > should create a domestic order 1`] = `
{
  "cubic_metre_volume": null,
  "customer_reference": "1337",
  "description": "Test",
  "hide_pickup_address": false,
  "kilogram_weight": "1.0",
  "labels": Any<Array>,
  "metadata": {
    "userId": 100,
  },
  "order_id": Any<String>,
  "order_url": Any<String>,
  "packaging_type": "box",
  "price": {
    "gross": {
      "amount": Any<Number>,
      "currency": "AUD",
    },
    "net": {
      "amount": Any<Number>,
      "currency": "AUD",
    },
    "tax": {
      "amount": Any<Number>,
      "currency": "AUD",
    },
  },
  "price_breakdown": {
    "base": {
      "amount": Any<Number>,
      "currency": "AUD",
    },
    "base_tax": {
      "amount": Any<Number>,
      "currency": "AUD",
    },
    "cover": {
      "amount": 0,
      "currency": "AUD",
    },
    "cover_tax": {
      "amount": 0,
      "currency": "AUD",
    },
    "discount": {
      "amount": 0,
      "currency": "AUD",
    },
    "discount_tax": {
      "amount": 0,
      "currency": "AUD",
    },
    "fuel_surcharge": {
      "amount": Any<Number>,
      "currency": "AUD",
    },
    "fuel_surcharge_tax": {
      "amount": Any<Number>,
      "currency": "AUD",
    },
  },
  "receiver": {
    "address": {
      "address_line1": "80 Wentworth Park Road",
      "address_line2": null,
      "country": "Australia",
      "postcode": "0800",
      "state_name": "NT",
      "suburb": "Darwin",
    },
    "contact": {
      "company": "Daily Planet",
      "email": "clarkissuper@dailyplanet.xyz",
      "name": "Clark Kent",
      "phone": null,
    },
    "instructions": "Signature on Delivery",
  },
  "route": {
    "description": "Sydney, Australia to Remote WA and NT, Australia",
    "type": "remote",
  },
  "scheduling": {
    "delivered_on": null,
    "estimated_delivery_date_maximum": Any<String>,
    "estimated_delivery_date_minimum": Any<String>,
    "is_cancellable": true,
    "picked_up_on": null,
    "pickup_date": Any<String>,
    "status": null,
  },
  "sender": {
    "address": {
      "address_line1": "123 Gotham Ln",
      "address_line2": null,
      "country": "Australia",
      "postcode": "2000",
      "state_name": "NSW",
      "suburb": "Sydney",
    },
    "contact": {
      "company": null,
      "email": Any<String>,
      "name": "Lex Luthor",
      "phone": null,
      "sendle_id": Any<String>,
    },
    "instructions": null,
  },
  "sendle_reference": Any<String>,
  "state": "Pickup",
  "tax_breakdown": {
    "gst": {
      "amount": Any<Number>,
      "currency": "AUD",
      "rate": 0.1,
    },
  },
  "tracking_url": Any<String>,
  "volume": null,
  "weight": {
    "units": "kg",
    "value": "1.0",
  },
}
`;

exports[`Orders > should create an international order 1`] = `
{
  "contents": {
    "country_of_origin": "China",
    "currency": "AUD",
    "description": "T-shirt",
    "value": "20.0",
  },
  "cubic_metre_volume": "0.01",
  "customer_reference": "SupBdayPressie",
  "description": "Kryptonite",
  "hide_pickup_address": false,
  "kilogram_weight": "1.0",
  "labels": Any<Array>,
  "metadata": {
    "your_data": "XYZ123",
  },
  "order_id": Any<String>,
  "order_url": Any<String>,
  "packaging_type": "box",
  "parcel_contents": [
    {
      "country_of_origin": "China",
      "currency": "AUD",
      "description": "T-shirt",
      "hs_code": "610910",
      "quantity": 1,
      "value": "20.0",
    },
  ],
  "price": {
    "gross": {
      "amount": Any<Number>,
      "currency": "AUD",
    },
    "net": {
      "amount": Any<Number>,
      "currency": "AUD",
    },
    "tax": {
      "amount": Any<Number>,
      "currency": "AUD",
    },
  },
  "price_breakdown": {
    "base": {
      "amount": Any<Number>,
      "currency": "AUD",
    },
    "base_tax": {
      "amount": Any<Number>,
      "currency": "AUD",
    },
    "cover": {
      "amount": 0,
      "currency": "AUD",
    },
    "cover_tax": {
      "amount": 0,
      "currency": "AUD",
    },
    "discount": {
      "amount": 0,
      "currency": "AUD",
    },
    "discount_tax": {
      "amount": 0,
      "currency": "AUD",
    },
    "fuel_surcharge": {
      "amount": Any<Number>,
      "currency": "AUD",
    },
    "fuel_surcharge_tax": {
      "amount": Any<Number>,
      "currency": "AUD",
    },
  },
  "receiver": {
    "address": {
      "address_line1": "445 Mount Eden Road",
      "address_line2": null,
      "country": "New Zealand",
      "postcode": "2025",
      "state_name": null,
      "suburb": "Auckland",
    },
    "contact": {
      "company": null,
      "email": "clarkissuper@dailyplanet.xyz",
      "name": "Clark Kent",
      "phone": null,
    },
    "instructions": "Give directly to Clark",
  },
  "route": {
    "description": "Sydney, Australia to New Zealand",
    "type": "export",
  },
  "scheduling": {
    "delivered_on": null,
    "estimated_delivery_date_maximum": Any<String>,
    "estimated_delivery_date_minimum": Any<String>,
    "is_cancellable": true,
    "picked_up_on": null,
    "pickup_date": Any<String>,
    "status": null,
  },
  "sender": {
    "address": {
      "address_line1": "123 Gotham Ln",
      "address_line2": null,
      "country": "Australia",
      "postcode": "2000",
      "state_name": "NSW",
      "suburb": "Sydney",
    },
    "contact": {
      "company": null,
      "email": Any<String>,
      "name": "Lex Luthor",
      "phone": "0412 345 678",
      "sendle_id": Any<String>,
    },
    "instructions": "Knock loudly",
  },
  "sendle_reference": Any<String>,
  "state": "Pickup",
  "tracking_url": Any<String>,
  "volume": {
    "units": "m3",
    "value": "0.01",
  },
  "weight": {
    "units": "kg",
    "value": "1.0",
  },
}
`;

exports[`Orders > should return tracking for my order 1`] = `
{
  "destination": {
    "country": "AU",
  },
  "origin": {
    "country": "AU",
  },
  "scheduling": {
    "delivered_on": null,
    "estimated_delivery_date_maximum": Any<String>,
    "estimated_delivery_date_minimum": Any<String>,
    "picked_up_on": null,
    "pickup_date": Any<String>,
    "status": null,
  },
  "state": "Pickup",
  "status": {
    "description": Any<String>,
    "last_changed_at": Any<String>,
  },
  "tracking_events": [],
}
`;
